Character encoding issues

I have a really strange bit of code I wrote a long time ago to test something.  I happened to cause a bug with this formatter where it fails to process the file.

Code to cause the error
```java
    public static void main(final String[] args) throws CharacterCodingException {
        final CharsetDecoder decoder = Charsets.ISO_8859_1.newDecoder();
        final CharsetEncoder encoder = Charsets.UTF_8.newEncoder();

        // Convert the byte array from starting inputEncoding into UCS2
        final byte[] bufferToConvert = "Hello World! £€".getBytes(Charsets.UTF_8);
        final CharBuffer cbuf = decoder.decode(ByteBuffer.wrap(bufferToConvert));

        // Convert the internal UCS2 representation into outputEncoding
        final ByteBuffer bbuf = encoder.encode(CharBuffer.wrap(cbuf));
        System.out.println(new String(bbuf.array(), 0, bbuf.limit(), Charsets.UTF_8));
    }
```

Here is the stack trace.

```
   at net.revelc.code.impsort.maven.plugin.AbstractImpSortMojo.fail (AbstractImpSortMojo.java:231)
    at net.revelc.code.impsort.maven.plugin.AbstractImpSortMojo.lambda$execute$3 (AbstractImpSortMojo.java:175)
    at java.util.stream.ReferencePipeline$3$1.accept (ReferencePipeline.java:193)
    at java.util.stream.ReferencePipeline$3$1.accept (ReferencePipeline.java:193)
    at java.util.stream.ForEachOps$ForEachOp$OfRef.accept (ForEachOps.java:184)
    at java.util.stream.ReferencePipeline$3$1.accept (ReferencePipeline.java:193)
    at java.util.Spliterators$ArraySpliterator.forEachRemaining (Spliterators.java:948)
    at java.util.stream.AbstractPipeline.copyInto (AbstractPipeline.java:481)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto (AbstractPipeline.java:471)
    at java.util.stream.ForEachOps$ForEachOp.evaluateSequential (ForEachOps.java:151)
    at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential (ForEachOps.java:174)
    at java.util.stream.AbstractPipeline.evaluate (AbstractPipeline.java:234)
    at java.util.stream.ReferencePipeline.forEach (ReferencePipeline.java:418)
    at java.util.stream.ReferencePipeline$7$1.accept (ReferencePipeline.java:270)
    at java.util.Spliterators$ArraySpliterator.forEachRemaining (Spliterators.java:948)
    at java.util.stream.AbstractPipeline.copyInto (AbstractPipeline.java:481)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto (AbstractPipeline.java:471)
    at java.util.stream.ReduceOps$ReduceTask.doLeaf (ReduceOps.java:747)
    at java.util.stream.ReduceOps$ReduceTask.doLeaf (ReduceOps.java:721)
    at java.util.stream.AbstractTask.compute (AbstractTask.java:316)
    at java.util.concurrent.CountedCompleter.exec (CountedCompleter.java:731)
    at java.util.concurrent.ForkJoinTask.doExec (ForkJoinTask.java:289)
    at java.util.concurrent.ForkJoinPool$WorkQueue.runTask (ForkJoinPool.java:1056)
    at java.util.concurrent.ForkJoinPool.runWorker (ForkJoinPool.java:1692)
    at java.util.concurrent.ForkJoinWorkerThread.run (ForkJoinWorkerThread.java:157)
Caused by: java.nio.charset.MalformedInputException: Input length = 1
    at java.nio.charset.CoderResult.throwException (CoderResult.java:281)
    at sun.nio.cs.StreamDecoder.implRead (StreamDecoder.java:339)
    at sun.nio.cs.StreamDecoder.read (StreamDecoder.java:178)
    at java.io.InputStreamReader.read (InputStreamReader.java:184)
    at java.io.BufferedReader.fill (BufferedReader.java:161)
    at java.io.BufferedReader.readLine (BufferedReader.java:324)
    at java.io.BufferedReader.readLine (BufferedReader.java:389)
    at java.nio.file.Files.readAllLines (Files.java:3205)
    at java.nio.file.Files.readAllLines (Files.java:3242)
    at net.revelc.code.impsort.ImpSort.parseFile (ImpSort.java:50)
    at net.revelc.code.impsort.maven.plugin.AbstractImpSortMojo.lambda$execute$3 (AbstractImpSortMojo.java:166)
    at java.util.stream.ReferencePipeline$3$1.accept (ReferencePipeline.java:193)
    at java.util.stream.ReferencePipeline$3$1.accept (ReferencePipeline.java:193)
    at java.util.stream.ForEachOps$ForEachOp$OfRef.accept (ForEachOps.java:184)
    at java.util.stream.ReferencePipeline$3$1.accept (ReferencePipeline.java:193)
    at java.util.Spliterators$ArraySpliterator.forEachRemaining (Spliterators.java:948)
    at java.util.stream.AbstractPipeline.copyInto (AbstractPipeline.java:481)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto (AbstractPipeline.java:471)
    at java.util.stream.ForEachOps$ForEachOp.evaluateSequential (ForEachOps.java:151)
    at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential (ForEachOps.java:174)
    at java.util.stream.AbstractPipeline.evaluate (AbstractPipeline.java:234)
    at java.util.stream.ReferencePipeline.forEach (ReferencePipeline.java:418)
    at java.util.stream.ReferencePipeline$7$1.accept (ReferencePipeline.java:270)
    at java.util.Spliterators$ArraySpliterator.forEachRemaining (Spliterators.java:948)
    at java.util.stream.AbstractPipeline.copyInto (AbstractPipeline.java:481)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto (AbstractPipeline.java:471)
    at java.util.stream.ReduceOps$ReduceTask.doLeaf (ReduceOps.java:747)
    at java.util.stream.ReduceOps$ReduceTask.doLeaf (ReduceOps.java:721)
    at java.util.stream.AbstractTask.compute (AbstractTask.java:316)
    at java.util.concurrent.CountedCompleter.exec (CountedCompleter.java:731)
    at java.util.concurrent.ForkJoinTask.doExec (ForkJoinTask.java:289)
    at java.util.concurrent.ForkJoinPool$WorkQueue.runTask (ForkJoinPool.java:1056)
    at java.util.concurrent.ForkJoinPool.runWorker (ForkJoinPool.java:1692)
    at java.util.concurrent.ForkJoinWorkerThread.run (ForkJoinWorkerThread.java:157)
```

--- a/src/main/java/net/revelc/code/impsort/ImpSort.java
+++ b/src/main/java/net/revelc/code/impsort/ImpSort.java
@@ -46,24 +46,25 @@ public class ImpSort {
   private static final Comparator<Node> BY_POSITION =
       (a, b) -> a.getBegin().get().compareTo(b.getBegin().get());
 
-  private final String sourceEncoding;
+  private final Charset sourceEncoding;
   private final Grouper grouper;
   private final boolean removeUnused;
 
-  public ImpSort(final String sourceEncoding, final Grouper grouper, final boolean removeUnused) {
+  public ImpSort(final Charset sourceEncoding, final Grouper grouper, final boolean removeUnused) {
     this.sourceEncoding = sourceEncoding;
     this.grouper = grouper;
     this.removeUnused = removeUnused;
   }
 
   public Result parseFile(final Path path) throws IOException {
-    List<String> fileLines = Files.readAllLines(path, Charset.forName(sourceEncoding));
+    List<String> fileLines = Files.readAllLines(path, sourceEncoding);
     CompilationUnit unit = JavaParser.parse(String.join("\n", fileLines));
     Position packagePosition =
         unit.getPackageDeclaration().map(p -> p.getEnd().get()).orElse(unit.getBegin().get());
     NodeList<ImportDeclaration> importDeclarations = unit.getImports();
     if (importDeclarations.isEmpty()) {
-      return new Result(path, fileLines, 0, fileLines.size(), "", "", Collections.emptyList());
+      return new Result(path, sourceEncoding, fileLines, 0, fileLines.size(), "", "",
+          Collections.emptyList());
     }
 
     // find orphaned comments before between package and last import
@@ -107,7 +108,8 @@ public class ImpSort {
       newSection += "\n";
     }
 
-    return new Result(path, fileLines, start, stop, originalSection, newSection, allImports);
+    return new Result(path, sourceEncoding, fileLines, start, stop, originalSection, newSection,
+        allImports);
   }
 
   // return imports, with associated comments, in order found in the file

--- a/src/main/java/net/revelc/code/impsort/ImpSort.java
+++ b/src/main/java/net/revelc/code/impsort/ImpSort.java
@@ -15,6 +15,7 @@
 package net.revelc.code.impsort;
 
 import java.io.IOException;
+import java.nio.charset.Charset;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.ArrayList;
@@ -45,16 +46,18 @@ public class ImpSort {
   private static final Comparator<Node> BY_POSITION =
       (a, b) -> a.getBegin().get().compareTo(b.getBegin().get());
 
+  private final String sourceEncoding;
   private final Grouper grouper;
   private final boolean removeUnused;
 
-  public ImpSort(final Grouper grouper, final boolean removeUnused) {
+  public ImpSort(final String sourceEncoding, final Grouper grouper, final boolean removeUnused) {
+    this.sourceEncoding = sourceEncoding;
     this.grouper = grouper;
     this.removeUnused = removeUnused;
   }
 
   public Result parseFile(final Path path) throws IOException {
-    List<String> fileLines = Files.readAllLines(path);
+    List<String> fileLines = Files.readAllLines(path, Charset.forName(sourceEncoding));
     CompilationUnit unit = JavaParser.parse(String.join("\n", fileLines));
     Position packagePosition =
         unit.getPackageDeclaration().map(p -> p.getEnd().get()).orElse(unit.getBegin().get());

--- a/src/main/java/net/revelc/code/impsort/Result.java
+++ b/src/main/java/net/revelc/code/impsort/Result.java
@@ -15,6 +15,7 @@
 package net.revelc.code.impsort;
 
 import java.io.IOException;
+import java.nio.charset.Charset;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.StandardCopyOption;
@@ -28,6 +29,7 @@ public class Result {
   private Boolean isSorted;
 
   private final Path path;
+  private final Charset sourceEncoding;
   private final String originalSection;
   private final String newSection;
   private final Collection<Import> allImports;
@@ -35,9 +37,10 @@ public class Result {
   private final int start;
   private final int stop;
 
-  Result(Path path, List<String> fileLines, int start, int stop, String originalSection,
-      String newSection, Collection<Import> allImports) {
+  Result(Path path, Charset sourceEncoding, List<String> fileLines, int start, int stop,
+      String originalSection, String newSection, Collection<Import> allImports) {
     this.path = path;
+    this.sourceEncoding = sourceEncoding;
     this.originalSection = originalSection;
     this.newSection = newSection;
     this.allImports = allImports;
@@ -62,7 +65,7 @@ public class Result {
   }
 
   public void saveSorted(Path destination) throws IOException {
-    if (isSorted) {
+    if (isSorted()) {
       if (!Files.isSameFile(path, destination)) {
         saveBackup(destination);
       }
@@ -79,7 +82,7 @@ public class Result {
       allLines.add(""); // restore blank line lost by split("\\n")
     }
     allLines.addAll(afterImports);
-    Files.write(destination, allLines);
+    Files.write(destination, allLines, sourceEncoding);
   }
 
 }

--- a/src/main/java/net/revelc/code/impsort/maven/plugin/AbstractImpSortMojo.java
+++ b/src/main/java/net/revelc/code/impsort/maven/plugin/AbstractImpSortMojo.java
@@ -16,6 +16,7 @@ package net.revelc.code.impsort.maven.plugin;
 
 import java.io.File;
 import java.io.IOException;
+import java.nio.charset.StandardCharsets;
 import java.nio.file.Path;
 import java.time.Duration;
 import java.util.concurrent.atomic.AtomicLong;
@@ -45,6 +46,9 @@ abstract class AbstractImpSortMojo extends AbstractMojo {
 
   @Parameter(defaultValue = "${plugin}", readonly = true)
   protected PluginDescriptor plugin;
+
+  @Parameter(defaultValue = "${project.build.sourceEncoding}", readonly = true)
+  protected String sourceEncoding = StandardCharsets.UTF_8.name();
 
   /**
    * Allows skipping execution of this plugin.
@@ -171,7 +175,7 @@ abstract class AbstractImpSortMojo extends AbstractMojo {
 
     // process all found files, and aggregate any failures
     Grouper grouper = new Grouper(groups, staticGroups, staticAfter, joinStaticWithNonStatic);
-    ImpSort impSort = new ImpSort(grouper, removeUnused);
+    ImpSort impSort = new ImpSort(sourceEncoding, grouper, removeUnused);
     AtomicLong numAlreadySorted = new AtomicLong(0);
     AtomicLong numProcessed = new AtomicLong(0);
 

--- a/src/main/java/net/revelc/code/impsort/maven/plugin/AbstractImpSortMojo.java
+++ b/src/main/java/net/revelc/code/impsort/maven/plugin/AbstractImpSortMojo.java
@@ -16,6 +16,7 @@ package net.revelc.code.impsort.maven.plugin;
 
 import java.io.File;
 import java.io.IOException;
+import java.nio.charset.Charset;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Path;
 import java.time.Duration;
@@ -175,7 +176,8 @@ abstract class AbstractImpSortMojo extends AbstractMojo {
 
     // process all found files, and aggregate any failures
     Grouper grouper = new Grouper(groups, staticGroups, staticAfter, joinStaticWithNonStatic);
-    ImpSort impSort = new ImpSort(sourceEncoding, grouper, removeUnused);
+    Charset encoding = Charset.forName(sourceEncoding);
+    ImpSort impSort = new ImpSort(encoding, grouper, removeUnused);
     AtomicLong numAlreadySorted = new AtomicLong(0);
     AtomicLong numProcessed = new AtomicLong(0);
 

--- a/src/test/java/net/revelc/code/impsort/ImpSortTest.java
+++ b/src/test/java/net/revelc/code/impsort/ImpSortTest.java
@@ -14,12 +14,15 @@
 
 package net.revelc.code.impsort;
 
+import static org.junit.Assert.assertArrayEquals;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
 
+import java.io.File;
 import java.io.IOException;
 import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.Arrays;
@@ -36,14 +39,14 @@ public class ImpSortTest {
   public void testSort() throws IOException {
     Path p = Paths.get(System.getProperty("user.dir"), "src", "it", "plugin-test", "src", "test",
         "java", "net", "revelc", "code", "imp", "PluginIT.java");
-    new ImpSort(StandardCharsets.UTF_8.name(), eclipseDefaults, false).parseFile(p);
+    new ImpSort(StandardCharsets.UTF_8, eclipseDefaults, false).parseFile(p);
   }
 
   @Test
   public void testUnused() throws IOException {
     Path p =
         Paths.get(System.getProperty("user.dir"), "src", "test", "resources", "UnusedImports.java");
-    Result result = new ImpSort(StandardCharsets.UTF_8.name(), eclipseDefaults, true).parseFile(p);
+    Result result = new ImpSort(StandardCharsets.UTF_8, eclipseDefaults, true).parseFile(p);
     Set<String> imports = new HashSet<>();
     for (Import i : result.getImports()) {
       imports.add(i.getImport());
@@ -75,8 +78,10 @@ public class ImpSortTest {
   public void testIso8859ForIssue3() throws IOException {
     Path p =
         Paths.get(System.getProperty("user.dir"), "src", "test", "resources", "Iso8859File.java");
-    Result result =
-        new ImpSort(StandardCharsets.ISO_8859_1.name(), eclipseDefaults, true).parseFile(p);
+    Result result = new ImpSort(StandardCharsets.ISO_8859_1, eclipseDefaults, true).parseFile(p);
     assertTrue(result.getImports().isEmpty());
+    Path output = File.createTempFile("impSort", null).toPath();
+    result.saveSorted(output);
+    assertArrayEquals(Files.readAllBytes(p), Files.readAllBytes(output));
   }
 }

--- a/src/test/java/net/revelc/code/impsort/ImpSortTest.java
+++ b/src/test/java/net/revelc/code/impsort/ImpSortTest.java
@@ -31,6 +31,8 @@ import java.util.Set;
 
 import org.junit.Test;
 
+import com.google.common.primitives.Bytes;
+
 public class ImpSortTest {
 
   private static Grouper eclipseDefaults = new Grouper("java.,javax.,org.,com.", "", false, false);
@@ -82,6 +84,10 @@ public class ImpSortTest {
     assertTrue(result.getImports().isEmpty());
     Path output = File.createTempFile("impSort", null).toPath();
     result.saveSorted(output);
-    assertArrayEquals(Files.readAllBytes(p), Files.readAllBytes(output));
+    byte[] testData = Files.readAllBytes(p);
+    // ensure expected ISO_8859_1 byte is present in test data, this defends against file being
+    // wrongly encoded if edited
+    assertTrue(Bytes.contains(testData, (byte) 0xe9));
+    assertArrayEquals(testData, Files.readAllBytes(output));
   }
 }

--- a/src/test/java/net/revelc/code/impsort/ImpSortTest.java
+++ b/src/test/java/net/revelc/code/impsort/ImpSortTest.java
@@ -19,6 +19,7 @@ import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
 
 import java.io.IOException;
+import java.nio.charset.StandardCharsets;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.Arrays;
@@ -35,14 +36,14 @@ public class ImpSortTest {
   public void testSort() throws IOException {
     Path p = Paths.get(System.getProperty("user.dir"), "src", "it", "plugin-test", "src", "test",
         "java", "net", "revelc", "code", "imp", "PluginIT.java");
-    new ImpSort(eclipseDefaults, false).parseFile(p);
+    new ImpSort(StandardCharsets.UTF_8.name(), eclipseDefaults, false).parseFile(p);
   }
 
   @Test
   public void testUnused() throws IOException {
     Path p =
         Paths.get(System.getProperty("user.dir"), "src", "test", "resources", "UnusedImports.java");
-    Result result = new ImpSort(eclipseDefaults, true).parseFile(p);
+    Result result = new ImpSort(StandardCharsets.UTF_8.name(), eclipseDefaults, true).parseFile(p);
     Set<String> imports = new HashSet<>();
     for (Import i : result.getImports()) {
       imports.add(i.getImport());
@@ -70,4 +71,12 @@ public class ImpSortTest {
         Grouper.parse(" b , * , ab ,com., java , a"));
   }
 
+  @Test
+  public void testIso8859ForIssue3() throws IOException {
+    Path p =
+        Paths.get(System.getProperty("user.dir"), "src", "test", "resources", "Iso8859File.java");
+    Result result =
+        new ImpSort(StandardCharsets.ISO_8859_1.name(), eclipseDefaults, true).parseFile(p);
+    assertTrue(result.getImports().isEmpty());
+  }
 }

--- a/src/test/resources/Iso8859File.java
+++ b/src/test/resources/Iso8859File.java
@@ -1,0 +1,19 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.revelc.code.imp;
+
+/** é */
+public class Iso8859File {
+}

